File: app/src/main/java/com/BlizzardArmory/connection/NetworkServices.java
package com.BlizzardArmory.connection;

import com.BlizzardArmory.UserInformation;
import com.BlizzardArmory.diablo.account.AccountInformation;
import com.BlizzardArmory.diablo.character.CharacterInformation;
import com.BlizzardArmory.diablo.item.SingleItem;
import com.BlizzardArmory.diablo.items.Items;
import com.BlizzardArmory.starcraft.Player;
import com.BlizzardArmory.starcraft.profile.Profile;
import com.BlizzardArmory.warcraft.account.Account;
import com.BlizzardArmory.warcraft.charactersummary.CharacterSummary;
import com.BlizzardArmory.warcraft.encounters.EncountersInformation;
import com.BlizzardArmory.warcraft.equipment.Equipment;
import com.BlizzardArmory.warcraft.media.Media;
import com.BlizzardArmory.warcraft.pvp.bracket.BracketStatistics;
import com.BlizzardArmory.warcraft.pvp.summary.PvPSummary;
import com.BlizzardArmory.warcraft.pvp.tiers.Tier;
import com.BlizzardArmory.warcraft.reputations.characterreputations.Reputation;
import com.BlizzardArmory.warcraft.statistic.Statistic;
import com.BlizzardArmory.warcraft.talents.Talents;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Url;

/**
 * The interface Network services.
 */
public interface NetworkServices {

    /**
     * Gets user info.
     *
     * @param accessToken the access token
     * @return the user info
     */
//User information
    @GET("oauth/userinfo")
    Call<UserInformation> getUserInfo(@Query("access_token") String accessToken);

    /**
     * Gets media.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the media
     */
//WoW Endpoints
    @GET("/profile/wow/character/{realm}/{charactername}/character-media")
    Call<Media> getMedia(@Path("charactername") String character,
                         @Path("realm") String realm,
                         @Query("namespace") String namespace,
                         @Query("locale") String locale,
                         @Query("access_token") String accessToken);

    /**
     * Gets dynamic equipment media.
     *
     * @param url         the url
     * @param locale      the locale
     * @param accessToken the access token
     * @return the dynamic equipment media
     */
    @GET
    Call<com.BlizzardArmory.warcraft.equipment.media.Media> getDynamicEquipmentMedia(@Url String url,
                                                                                     @Query("locale") String locale,
                                                                                     @Query("access_token") String accessToken);

    /**
     * Gets encounters.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the encounters
     */
    @GET("profile/wow/character/{realm}/{charactername}/encounters/raids")
    Call<EncountersInformation> getEncounters(@Path("charactername") String character,
                                              @Path("realm") String realm,
                                              @Query("namespace") String namespace,
                                              @Query("locale") String locale,
                                              @Query("access_token") String accessToken);

    /**
     * Gets equipped items.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the equipped items
     */
    @GET("profile/wow/character/{realm}/{charactername}/equipment")
    Call<Equipment> getEquippedItems(@Path("charactername") String character,
                                     @Path("realm") String realm,
                                     @Query("namespace") String namespace,
                                     @Query("locale") String locale,
                                     @Query("access_token") String accessToken);

    /**
     * Gets stats.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the stats
     */
    @GET("profile/wow/character/{realm}/{charactername}/statistics")
    Call<Statistic> getStats(@Path("charactername") String character,
                             @Path("realm") String realm,
                             @Query("namespace") String namespace,
                             @Query("locale") String locale,
                             @Query("access_token") String accessToken);

    /**
     * Gets specs.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the specs
     */
    @GET("profile/wow/character/{realm}/{charactername}/specializations")
    Call<Talents> getSpecs(@Path("charactername") String character,
                           @Path("realm") String realm,
                           @Query("namespace") String namespace,
                           @Query("locale") String locale,
                           @Query("access_token") String accessToken);


    /**
     * Gets character.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the character
     */
    @GET("profile/wow/character/{realm}/{charactername}")
    Call<CharacterSummary> getCharacter(@Path("charactername") String character,
                                        @Path("realm") String realm,
                                        @Query("namespace") String namespace,
                                        @Query("locale") String locale,
                                        @Query("access_token") String accessToken);

    /**
     * Gets account.
     *
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the account
     */
    @GET("profile/user/wow")
    Call<Account> getAccount(@Query("namespace") String namespace,
                             @Query("locale") String locale,
                             @Query("access_token") String accessToken);

    /**
     * Gets pv p summary.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the pv p summary
     */
    @GET("profile/wow/character/{realm}/{charactername}/pvp-summary")
    Call<PvPSummary> getPvPSummary(@Path("charactername") String character,
                                   @Path("realm") String realm,
                                   @Query("namespace") String namespace,
                                   @Query("locale") String locale,
                                   @Query("access_token") String accessToken);

    /**
     * Gets pv p brackets.
     *
     * @param character   the character
     * @param realm       the realm
     * @param bracket     the bracket
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the pv p brackets
     */
    @GET("profile/wow/character/{realm}/{charactername}/pvp-bracket/{BRACKET}")
    Call<BracketStatistics> getPvPBrackets(@Path("charactername") String character,
                                           @Path("realm") String realm,
                                           @Path("BRACKET") String bracket,
                                           @Query("namespace") String namespace,
                                           @Query("locale") String locale,
                                           @Query("access_token") String accessToken);

    /**
     * Gets dynamic tier.
     *
     * @param url         the url
     * @param locale      the locale
     * @param accessToken the access token
     * @return the dynamic tier
     */
    @GET
    Call<Tier> getDynamicTier(@Url String url,
                              @Query("locale") String locale,
                              @Query("access_token") String accessToken);

    /**
     * Gets reputations.
     *
     * @param character   the character
     * @param realm       the realm
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the reputations
     */
    @GET("profile/wow/character/{realm}/{charactername}/reputations")
    Call<Reputation> getReputations(@Path("charactername") String character,
                                    @Path("realm") String realm,
                                    @Query("namespace") String namespace,
                                    @Query("locale") String locale,
                                    @Query("access_token") String accessToken);

    /**
     * Gets d 3 profile.
     *
     * @param battletag   the battletag
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the d 3 profile
     */
//D3 Endpoints
    @GET("d3/profile/{battletag}/")
    Call<AccountInformation> getD3Profile(@Path("battletag") String battletag,
                                          @Query("namespace") String namespace,
                                          @Query("locale") String locale,
                                          @Query("access_token") String accessToken);

    /**
     * Gets d 3 hero.
     *
     * @param battletag   the battletag
     * @param id          the id
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the d 3 hero
     */
    @GET("d3/profile/{battletag}/hero/{id}")
    Call<CharacterInformation> getD3Hero(@Path("battletag") String battletag,
                                         @Path("id") long id,
                                         @Query("namespace") String namespace,
                                         @Query("locale") String locale,
                                         @Query("access_token") String accessToken);

    /**
     * Gets hero items.
     *
     * @param battletag   the battletag
     * @param id          the id
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the hero items
     */
    @GET("d3/profile/{battletag}/hero/{id}/items")
    Call<Items> getHeroItems(@Path("battletag") String battletag,
                             @Path("id") long id,
                             @Query("namespace") String namespace,
                             @Query("locale") String locale,
                             @Query("access_token") String accessToken);

    /**
     * Gets item.
     *
     * @param item        the item
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the item
     */
    @GET("d3/data/{item}")
    Call<SingleItem> getItem(@Path("item") String item,
                             @Query("namespace") String namespace,
                             @Query("locale") String locale,
                             @Query("access_token") String accessToken);

    /**
     * Gets sc 2 player.
     *
     * @param id          the id
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the sc 2 player
     */
//Sc2 endpoints
    @GET("sc2/player/{id}")
    Call<List<Player>> getSc2Player(@Path("id") String id,
                                    @Query("namespace") String namespace,
                                    @Query("locale") String locale,
                                    @Query("access_token") String accessToken);

    /**
     * Gets sc 2 profile.
     *
     * @param regionId    the region id
     * @param realmId     the realm id
     * @param profileId   the profile id
     * @param namespace   the namespace
     * @param locale      the locale
     * @param accessToken the access token
     * @return the sc 2 profile
     */
    @GET("sc2/profile/{region_id}/{realm_id}/{profile_id}")
    Call<Profile> getSc2Profile(@Path("region_id") int regionId,
                                @Path("realm_id") int realmId,
                                @Path("profile_id") String profileId,
                                @Query("namespace") String namespace,
                                @Query("locale") String locale,
                                @Query("access_token") String accessToken);

    /**
     * Gets ow profile.
     *
     * @param url the url
     * @return the ow profile
     */
//Overwatch endpoint
    @GET
    Call<com.BlizzardArmory.overwatch.Profile> getOWProfile(@Url String url);

}

File: app/src/main/java/com/BlizzardArmory/connection/NetworkServices.kt
package com.BlizzardArmory.connection

import com.BlizzardArmory.model.UserInformation
import com.BlizzardArmory.model.diablo.account.AccountInformation
import com.BlizzardArmory.model.diablo.character.CharacterInformation
import com.BlizzardArmory.model.diablo.data.common.Leaderboard
import com.BlizzardArmory.model.diablo.data.eras.era.Era
import com.BlizzardArmory.model.diablo.data.eras.index.EraIndex
import com.BlizzardArmory.model.diablo.data.seasons.index.SeasonIndex
import com.BlizzardArmory.model.diablo.data.seasons.season.Season
import com.BlizzardArmory.model.diablo.item.SingleItem
import com.BlizzardArmory.model.diablo.items.Items
import com.BlizzardArmory.model.starcraft.Player
import com.BlizzardArmory.model.starcraft.profile.Profile
import com.BlizzardArmory.model.warcraft.account.Account
import com.BlizzardArmory.model.warcraft.achievements.categories.Categories
import com.BlizzardArmory.model.warcraft.achievements.characterachievements.Achievements
import com.BlizzardArmory.model.warcraft.achievements.custom.DetailedAchievements
import com.BlizzardArmory.model.warcraft.charactersummary.CharacterSummary
import com.BlizzardArmory.model.warcraft.covenant.character.soulbind.SoulbindInformation
import com.BlizzardArmory.model.warcraft.covenant.covenant.custom.CovenantSpells
import com.BlizzardArmory.model.warcraft.encounters.EncountersInformation
import com.BlizzardArmory.model.warcraft.equipment.Equipment
import com.BlizzardArmory.model.warcraft.media.Media
import com.BlizzardArmory.model.warcraft.pvp.bracket.BracketStatistics
import com.BlizzardArmory.model.warcraft.pvp.summary.PvPSummary
import com.BlizzardArmory.model.warcraft.pvp.tiers.Tier
import com.BlizzardArmory.model.warcraft.realm.Realms
import com.BlizzardArmory.model.warcraft.reputations.characterreputations.Reputation
import com.BlizzardArmory.model.warcraft.reputations.custom.ReputationPlusParentInfo
import com.BlizzardArmory.model.warcraft.statistic.Statistic
import com.BlizzardArmory.model.warcraft.talents.Talents
import com.BlizzardArmory.model.warcraft.talents.TalentsIcons
import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Url

interface NetworkServices {

    @GET("oauth/token")
    fun getAccessToken(@Query("code") clientId: String?,
                       @Query("region") region: String?,
                       @Query("redirect_uri") redirectUri: String?): Call<com.BlizzardArmory.connection.oauth.TokenResponse>

    //User information
    @GET("oauth/userinfo")
    fun getUserInfo(@Query("token") accessToken: String?,
                    @Query("region") region: String?): Call<UserInformation>

    @GET("/data/wow/realm/index")
    fun getRealmIndex(@Query("region") region: String,
                      @Query("namespace") namespace: String,
                      @Query("locale") locale: String,
                      @Query("token") accessToken: String?): Call<Realms>

    //WoW Endpoints
    @GET("/profile/wow/character/{realm}/{character}/character-media")
    fun getMedia(@Path("character", encoded = true) character: String?,
                 @Path("realm", encoded = true) realm: String?,
                 @Query("locale") locale: String?,
                 @Query("region") region: String?,
                 @Query("token") accessToken: String?): Call<Media>

    @GET
    fun getDynamicEquipmentMedia(@Url url: String?,
                                 @Query("locale") locale: String?,
                                 @Query("region") region: String?): Call<com.BlizzardArmory.model.warcraft.equipment.media.Media>

    @GET
    fun getAchievementCategories(@Url url: String?): Call<Categories>

    @GET
    fun getAllAchievements(@Url url: String?): Call<DetailedAchievements>

    @GET
    fun getTalentsWithIcon(@Url url: String?): Call<List<TalentsIcons>>

    @GET
    fun getCovenantSpells(@Url url: String): Call<List<CovenantSpells>>

    @GET
    fun getReputationPlusParentInfo(@Url url: String): Call<List<ReputationPlusParentInfo>>

    @GET("/profile/wow/character/{realm}/{character}/achievements")
    fun getCharacterAchievements(@Path("character", encoded = true) character: String?,
                                 @Path("realm", encoded = true) realm: String?,
                                 @Query("locale") locale: String?,
                                 @Query("region") region: String?,
                                 @Query("token") accessToken: String?): Call<Achievements>

    @GET("profile/wow/character/{realm}/{character}/encounters/raids")
    fun getEncounters(@Path("character", encoded = true) character: String?,
                      @Path("realm", encoded = true) realm: String?,
                      @Query("locale") locale: String?,
                      @Query("region") region: String?,
                      @Query("token") accessToken: String?): Call<EncountersInformation>

    @GET("profile/wow/character/{realm}/{character}/equipment")
    fun getEquippedItems(@Path("character", encoded = true) character: String?,
                         @Path("realm", encoded = true) realm: String?,
                         @Query("locale") locale: String?,
                         @Query("region") region: String?,
                         @Query("token") accessToken: String?): Call<Equipment>

    @GET("profile/wow/character/{realm}/{character}/statistics")
    fun getStats(@Path("character", encoded = true) character: String?,
                 @Path("realm", encoded = true) realm: String?,
                 @Query("locale") locale: String?,
                 @Query("region") region: String?,
                 @Query("token") accessToken: String?): Call<Statistic>

    @GET("profile/wow/character/{realm}/{character}/specializations")
    fun getSpecs(@Path("character", encoded = true) character: String?,
                 @Path("realm", encoded = true) realm: String?,
                 @Query("locale") locale: String?,
                 @Query("region") region: String?,
                 @Query("token") accessToken: String?): Call<Talents>


    @GET("profile/wow/character/{realm}/{character}")
    fun getCharacter(@Path("character", encoded = true) character: String?,
                     @Path("realm", encoded = true) realm: String?,
                     @Query("locale") locale: String?,
                     @Query("region") region: String?,
                     @Query("token") accessToken: String?): Call<CharacterSummary>

    @GET("profile/user/wow")
    fun getAccount(@Query("locale") locale: String?,
                   @Query("region") region: String?,
                   @Query("token") accessToken: String?): Call<Account>

    @GET("profile/wow/character/{realm}/{character}/pvp-summary")
    fun getPvPSummary(@Path("character", encoded = true) character: String?,
                      @Path("realm", encoded = true) realm: String?,
                      @Query("locale") locale: String?,
                      @Query("region") region: String?,
                      @Query("token") accessToken: String?): Call<PvPSummary>

    @GET("profile/wow/character/{realm}/{character}/pvp-bracket/{BRACKET}")
    fun getPvPBrackets(@Path("character", encoded = true) character: String?,
                       @Path("realm", encoded = true) realm: String?,
                       @Path("BRACKET") bracket: String?,
                       @Query("locale") locale: String?,
                       @Query("region") region: String?,
                       @Query("token") accessToken: String?): Call<BracketStatistics>

    @GET
    fun getDynamicTier(@Url url: String?,
                       @Query("region") region: String?,
                       @Query("locale") locale: String?): Call<Tier>

    @GET("profile/wow/character/{realm}/{character}/reputations")
    fun getReputations(@Path("character", encoded = true) character: String?,
                       @Path("realm", encoded = true) realm: String?,
                       @Query("locale") locale: String?,
                       @Query("region") region: String?,
                       @Query("token") accessToken: String?): Call<Reputation>

    @GET("/profile/wow/character/{realm}/{character}/soulbinds")
    fun getSoulbinds(@Path("character", encoded = true) character: String?,
                     @Path("realm", encoded = true) realm: String?,
                     @Query("locale") locale: String?,
                     @Query("region") region: String?,
                     @Query("token") accessToken: String?): Call<SoulbindInformation>

    //D3 Endpoints
    @GET("d3/profile/{battletag}/")
    fun getD3Profile(@Path("battletag") battletag: String?,
                     @Query("locale") locale: String?,
                     @Query("region") region: String?,
                     @Query("token") accessToken: String?): Call<AccountInformation>

    @GET("d3/profile/{battletag}/hero/{id}")
    fun getD3Hero(@Path("battletag") battletag: String?,
                  @Path("id") id: Long,
                  @Query("locale") locale: String?,
                  @Query("region") region: String?,
                  @Query("token") accessToken: String?): Call<CharacterInformation>

    @GET("d3/profile/{battletag}/hero/{id}/items")
    fun getHeroItems(@Path("battletag") battletag: String?,
                     @Path("id") id: Long,
                     @Query("locale") locale: String?,
                     @Query("region") region: String?,
                     @Query("token") accessToken: String?): Call<Items>

    @GET("d3/data/item/{slug}")
    fun getItem(@Path("slug", encoded = true) slug: String?,
                @Query("region") region: String?,
                @Query("locale") locale: String?): Call<SingleItem>

    @GET("data/d3/season/")
    fun getSeasonIndex(@Query("locale") locale: String,
                       @Query("region") region: String): Call<SeasonIndex>

    @GET("data/d3/season/{id}")
    fun getSeasonById(@Path("id") id: Int,
                      @Query("locale") locale: String,
                      @Query("region") region: String): Call<Season>

    @GET("data/d3/season/{id}/leaderboard/{leaderboard}")
    fun getSeasonLeaderboard(@Path("id") id: Int,
                             @Path("leaderboard") leaderboard: String,
                             @Query("locale") locale: String,
                             @Query("region") region: String): Call<Leaderboard>

    @GET("data/d3/era/")
    fun getEraIndex(@Query("locale") locale: String,
                    @Query("region") region: String): Call<EraIndex>

    @GET("data/d3/era/{id}")
    fun getEraById(@Path("id") id: Int,
                   @Query("locale") locale: String,
                   @Query("region") region: String): Call<Era>

    @GET("data/d3/era/{id}/leaderboard/{leaderboard}")
    fun getEraLeaderboard(@Path("id") id: Int,
                          @Path("leaderboard") leaderboard: String,
                          @Query("locale") locale: String,
                          @Query("region") region: String): Call<Leaderboard>

    //Sc2 endpoints
    @GET("sc2/player/{id}")
    fun getSc2Player(@Path("id") id: String?,
                     @Query("locale") locale: String?,
                     @Query("region") region: String?,
                     @Query("token") accessToken: String?): Call<List<Player>>

    @GET("sc2/profile/{region_id}/{realm_id}/{profile_id}")
    fun getSc2Profile(@Path("region_id") regionId: String,
                      @Path("realm_id") realmId: Int,
                      @Path("profile_id") profileId: String?,
                      @Query("locale") locale: String?,
                      @Query("region") region: String?,
                      @Query("token") accessToken: String?): Call<Profile>

    //Overwatch endpoint
    @GET
    fun getOWProfile(@Url url: String?): Call<com.BlizzardArmory.model.overwatch.Profile>
}